### PR TITLE
feat(medusa-file-s3): Support for any s3 compatible service

### PR DESCRIPTION
**Pull request by ashutoshpw**
Users can pass the config options in the plugin so that all the parameters could be directly passed to the aws-sdk library. By using this approach, they can also do many other things apart from using just any other s3 compatible service.

### CODE REVIEW

File: packages/medusa-file-s3/src/services/s3.js [hunk 1]
```diff
@@ -3,28 +3,23 @@ import aws from "aws-sdk"
...
-    this.accessKeyId_ = options.access_key_id
-    this.secretAccessKey_ = options.secret_access_key
-    this.region_ = options.region
-    this.endpoint_ = options.endpoint
+
+    // Remove parameters not supported by aws-sdk
+    delete options.bucket
+
+    this.awsOptions_ = options
   }
```
Comment (srindom) [inline packages/medusa-file-s3/src/services/s3.js:15]: todo: we need this to remain backward compatible for anyone who might be using the plugin in its current state - can I ask you to add a "mapper" that takes the existing options and maps them to the correct corresponding awsOptions?

**Review by srindom — COMMENTED** [review]
Thank you for the contribution this is an excellent suggestion! Would be great if you could address my comment below :)

**Comment by ashutoshpw** [comment]
@srindom I have rebased it with `develop` branch. Should I do anything else for this to merge with develop?

**Review by shahednasser — CHANGES_REQUESTED** [review]
Thank you @ashutoshpw for your contribution for the documentation as well! To make sure people know that the S3 plugin can be used with S3 compatible services, can you add a note at the beginning of the documentation (after "In this document...") indicating that? Can be something like this:

```
:::note

This plugin is mainly created for S3, but can support S3 compatible services.

:::
```

**Comment by ashutoshpw** [comment]
@shahednasser docs are updated now for this one.

**Review by shahednasser — APPROVED** [review]
Docs look good to me 😄

File: packages/medusa-file-s3/src/services/s3.js (hunk 1 above)
Comment (olivermrbl) [reply packages/medusa-file-s3/src/services/s3.js:25]: @srindom Are you on board with the latest changes?

**Comment by RegisHubelia** [comment]
Hey, any news for this PR?  Been there for quite some time and it's overdue...

Thanks!

**Comment by Charlio99** [comment]
> @olivermrbl, All good - ended up modifying the current plugin so it works with S3 in general. But as this PR is coming soon, I'll simply go back to the plugin once it is merged.

Hey @olivermrbl! Do you mind sharing how did you do it? I need to test S3 integration with a non-AWS provider. Thanks!

**Comment by RegisHubelia** [comment]
yhea sure - here you go, didn't reinvent the wheel here, works with backblaze, didn't actually tried with AWS.  The public-read will give you trouble - I need to fix that as it's not the same depending on the S3 provider - but a private bucket on backblaze works fine.

`import fs from "fs"
import aws from "aws-sdk"
import { AbstractFileService } from "@medusajs/medusa"
import stream from "stream"

class S3Service extends AbstractFileService {
    // eslint-disable-next-line no-empty-pattern
    constructor({}, options) {
        super({}, options)
        options.isProtected = true
        this.bucket_ = process.env.S3_BUCKET
        this.s3Url_ = process.env.S3_URL
        this.accessKeyId_ = process.env.S3_ACCESS_KEY_ID
        this.secretAccessKey_ = process.env.S3_SECRET_ACCESS_KEY
        this.region_ = process.env.S3_REGION
        this.endpoint_ = process.env.S3_ENDPOINT || process.env.S3_URL
        this.s3_ = new aws.S3({endpoint: this.endpoint_, accessKeyId: this.accessKeyId_, secretAccessKey: this.secretAccessKey_})
    }

    upload(file) {
        this.updateAwsConfig()

        return this.uploadFile(file)
    }

    uploadProtected(file) {
        this.updateAwsConfig()

        return this.uploadFile(file, { acl: "private" })
    }

    uploadFile(file, options = { isProtected: false, acl: undefined }) {
        const s3 = this.s3_
        const params = {
            ACL: options.acl ?? (options.isProtected ? "private" : "public-read"),
            Bucket: this.bucket_,
            Body: fs.createReadStream(file.path),
            Key: `${file.originalname}`
        }

        return new Promise((resolve, reject) => {
            s3.upload(params, (err, data) => {
                if (err) {
                    reject(err)
                    return
                }

                resolve({ url: data.Location, key: data.Key })
            })
        })
    }

    async delete(file) {
        this.updateAwsConfig()

        const s3 = this.s3_
        const params = {
            Bucket: this.bucket_,
            Key: `${file}`,
        }

        return new Promise((resolve, reject) => {
            s3.deleteObject(params, (err, data) => {
                if (err) {
                    reject(err)
                    return
                }
                resolve(data)
            })
        })
    }

    async getUploadStreamDescriptor(fileData) {
        this.updateAwsConfig()

        const pass = new stream.PassThrough()

        const fileKey = `${fileData.name}.${fileData.ext}`
        const params = {
            ACL: fileData.acl ?? "private",
            Bucket: this.bucket_,
            Body: pass,
            Key: fileKey,
        }

        const s3 = this.s3_
        return {
            writeStream: pass,
            promise: s3.upload(params).promise(),
            url: `${this.s3Url_}/${fileKey}`,
            fileKey,
        }
    }

    async getDownloadStream(fileData) {
        this.updateAwsConfig()

        const s3 = this.s3_

        const params = {
            Bucket: this.bucket_,
            Key: `${fileData.fileKey}`,
        }

        return s3.getObject(params).createReadStream()
    }

    async getPresignedDownloadUrl(fileData) {
        this.updateAwsConfig({
            signatureVersion: "v4",
        })

        const s3 = this.s3_

        const params = {
            Bucket: this.bucket_,
            Key: `${fileData.fileKey}`,
            Expires: 320
        }

        return await s3.getSignedUrlPromise("getObject", params)
    }

    updateAwsConfig(additionalConfiguration = {}) {
        aws.config.setPromisesDependency(null)
        aws.config.update(
          {
              accessKeyId: this.accessKeyId_,
              secretAccessKey: this.secretAccessKey_,
              region: this.region_,
              endpoint: this.endpoint_,
              ...additionalConfiguration,
          },
          true
        )
    }
}

export default S3Service
`

**Comment by Charlio99** [comment]
@RegisHubelia I'm getting: Malformed Access Key Id

**Comment by RegisHubelia** [comment]
> @RegisHubelia I'm getting: Malformed Access Key Id. Did this happened to you? If so, how did you solve it?

I have not...  What S3 provider are you using?  Did you set the env variables?

S3_BUCKET=thebucket
S3_REGION=us-west-002
S3_ACCESS_KEY_ID=theaccesskeyid
S3_SECRET_ACCESS_KEY=thesecretaccesskey
S3_URL=https://s3.us-west-002.backblazeb2.com

**Comment by Charlio99** [comment]
> > @RegisHubelia I'm getting: Malformed Access Key Id. Did this happened to you? If so, how did you solve it?
> 
> I have not... What S3 provider are you using? Did you set the env variables?
> 
> S3_BUCKET=thebucket S3_REGION=us-west-002 S3_ACCESS_KEY_ID=theaccesskeyid S3_SECRET_ACCESS_KEY=thesecretaccesskey S3_URL=https://s3.us-west-002.backblazeb2.com

@RegisHubelia I think I found the issue, the package is not being patched, I'm using `npx patch-package medusa-file-s3`, and it doesn't seem to be working. How did you "patched" the package?

**Comment by RegisHubelia** [comment]
@Charlio99, you'll need to create a service and not use the actual package included with medusa.  You can have a look here:
https://docs.medusajs.com/advanced/backend/services/create-service/

Basically, created a file services/s3.js and paste the code I sent you.

**Comment by Charlio99** [comment]
> @Charlio99, you'll need to create a service and not use the actual package included with medusa. You can have a look here: https://docs.medusajs.com/advanced/backend/services/create-service/
> 
> Basically, created a file services/s3.js and paste the code I sent you.

@RegisHubelia that did the trick for backblaze, now I'm trying to get it to work with [Storj DCS](https://www.storj.io/) but it's not working, no errors are shown in UI or logs, it just says "succesful" as if the image was uploaded

**Comment by RegisHubelia** [comment]
It will need debugging...  I'm not familiar with storj - but came across this:
https://www.storj.io/blog/what-is-s3-compatibility

**Comment by Charlio99** [comment]
> It will need debugging... I'm not familiar with storj - but came across this: https://www.storj.io/blog/what-is-s3-compatibility

I know, but it doesn't work out of the box like backblaze with the code you sent

**Comment by riqwan** [comment]
Hey, thanks for the PR! Since v2 brought a lot of architectural and API changes on the backend, we will be closing this ticket.

**Pull request closed without merging**